### ISSUE TITLE
CleanUp of New GraphToGpu Optimizer

### ISSUE THREAD

**Issue by Sentient07** [Theano/Theano#4801]
The cleanup of PR #4570.
Tasks :

- [ ]  Correctly handle HostFromGpu nodes

- [ ]  Remove the special case for Alloc / AllocEmpty from the new optimizer, and register these lifters in an "out2in" topooptimizer

- [ ] Replace op_lifter with out2in

- [ ]  Factor out caching of Op instances

- [ ] Check if all the optimizers used by op_lifter is being used by op_lifter_topo
 Implement backward pass.

- [ ] Remove op_lifter from EquilibriumOptimizer

- [ ] Check the abnormal pattern of HostFromGpu and GpuFromHost combination in `local_cut_gpu_transfers`

**Comment by Sentient07** [comment]
The first four are being done in #4672 . The 5th and 6th could be done with the new `register_useless` optimizer.  The sixth is being done in a different branch which i haven't pushed (Haven't identified how the pattern, HostFromGpu(GpuFromHost(HostFromGpu(Variable)))) is introduced.  The seventh is being done in #4728

**Comment by nouiz** [comment]
We should not replace op_lifter with out2in. out2in would be a new optimization pass after GraphToGPU. But in the lastest profile, I don't think it will help. But this would help fix the Alloc, AllocEmpty special case.

Mostly out2in is a way to implement the back-ward pass. I'll re-work the description.

**Comment by nouiz** [comment]
I did that. I don't understand this one:
 
- Check if all the optimizers used by op_lifter is being used by op_lifter_topo

**Comment by nouiz** [comment]
also, register_useless can't help for the back-ward pass and the op_lifter.
register_useless is before we move to the GPU.

**Comment by Sentient07** [comment]
> We should not replace op_lifter with out2in. out2in would be a new optimization pass after GraphToGPU. 

What i meant was replace that from EquilibriumOptimizer as TopoOptimizer with out2_in/

> Check if all the optimizers used by op_lifter is being used by op_lifter_topo

**Comment by Sentient07** [comment]
> also, register_useless can't help for the back-ward pass and the op_lifter.
> register_useless is before we move to the GPU.

Yeah, we'd need another optimizer that is like register_useless for the `op_lifter` optimizers, right ?
